README.md
# CDK-based Keycloak Cluster on ECS

This CDK construct allows you to spin up a high availability Keycloak cluster on AWS ECS using Fargate capacity.

* [API Reference](https://github.com/wheatstalk/cdk-ecs-keycloak/blob/master/API.md)
* [Changelog](https://github.com/wheatstalk/cdk-ecs-keycloak/blob/master/CHANGELOG.md)

## What is Keycloak?

Keycloak is an open-source identity and access management server. With it, you can add authentication to your apps and secure services with little fuss.

## But why Keycloak on ECS over Cognito?

- Account linking doesn't break down in Keycloak when the user signed up with a social account and now wants to set a password
- User-oriented account management console
- More user federation, identity brokering, and social login features
- Store and access the user's federated access tokens on other services

## Examples

### All defaults

The simplest example shows all defaults below.

```ts
import * as keycloak from '@wheatstalk/cdk-ecs-keycloak';
import * as cdk from '@aws-cdk/core';

// Create a CDK app and a stack
const app = new cdk.App();
const stack = new cdk.Stack(app, 'Keycloak-test');

// Add a Keycloak cluster to the stack
new keycloak.KeycloakCluster(stack, 'Keycloak');
```

The initial deployment takes a few minutes, after which a URL to Keycloak will be shown in the terminal. After a few minutes of settling, Keycloak will be available at that URL.

By default, all the following are created:

- A VPC (public and private subnet with nat gateways)
- An Internet-facing load balancer serving HTTP (not HTTPS) traffic
- An ECS cluster with Keycloak running on it
- An Aurora Serverless MySQL cluster

### Auto-scaling with internal HTTPS and Keycloak configuration

To demonstrate more functionality, here is an example cluster that does the following:

- Requests Fargate tasks sized 0.5vCPU / 1GB RAM
- Controls ECS deployment min and max health settings
- Sets up Infiniscan clustering by increasing cache owner count
- Publishes an Application Load Balancer with internal HTTPS
- The Application Load Balancer in this example upgrades HTTP connections to HTTPS
- Sets up auto-scaling by interacting directly with the ECS service

```ts
// Create a Keycloak cluster on Fargate
const keycloakCluster = new keycloak.KeycloakCluster(this, 'Keycloak', {
  // Fargate task sizes
  cpu: 512,
  memoryLimitMiB: 1024,
  // Service options
  minHealthyPercent: 50,
  maxHealthyPercent: 200,
  keycloak: {
    // Set distributed inficaches owners to two
    cacheOwnersCount: 2,
  },
  // Use an HTTPS load balancer with internal HTTPS from the load balancer to Keycloak.
  httpsPortPublisher: keycloak.PortPublisher.httpsAlb({
    certificates: [certificate],
    // Redirect HTTP traffic to HTTPS
    upgradeHttp: true,
  }),
});

// Auto-scale the service
const autoScaling = keycloakCluster.service.autoScaleTaskCount({
  maxCapacity: 5,
  minCapacity: 3,
});

autoScaling.scaleOnCpuUtilization('Target40', {
  targetUtilizationPercent: 40,
  scaleInCooldown: cdk.Duration.minutes(30),
  scaleOutCooldown: cdk.Duration.minutes(10),
});
```

### Use a database instance

You may opt to use a database instance instead of an Aurora Serverless cluster.

```ts
new keycloak.KeycloakCluster(this, 'Keycloak', {
  databaseProvider: keycloak.DatabaseProvider.databaseInstance({
    engine: rds.DatabaseInstanceEngine.mysql({
      version: rds.MysqlEngineVersion.VER_5_7,
    }),
    instanceType: ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.MICRO),
  }),
});
```

### Use Postgres instead of MySQL

```ts
new keycloak.KeycloakCluster(this, 'Keycloak', {
  databaseProvider: keycloak.DatabaseProvider.databaseInstance({
    engine: rds.DatabaseInstanceEngine.postgres({
      version: rds.PostgresEngineVersion.VER_11_9,
    }),
    instanceType: ec2.InstanceType.of(ec2.InstanceClass.T2, ec2.InstanceSize.MICRO),
  }),
});
```

### Provide your VPC, Database and Application Load Balancer

You may provide resources you've created, such as VPCs, Clusters, CloudMap
namespaces, databases and load balancers.

In the following for example we re-use a VPC, database and Application Load
Balancer listener. As a result the Cluster construct will:

- Create a new ECS cluster in the given VPC
- Use the given database information
- Create an ingress rule in the database's security group
- Publish its HTTP port on the given load balancer listener

```ts
// Your resources
const vpc = new ec2.Vpc(...);
const rdsDb = new rds.DatabaseInstance(...);
const loadBalancer = new elbv2.ApplicationLoadBalancer(...);
const listener = loadBalancer.addListener(...);

new keycloak.KeycloakCluster(this, 'Keycloak', {
  // Provide an existing VPC so the cluster and database can opt to reuse it
  vpcProvider: keycloak.VpcProvider.fromVpc(vpc),
  // Bring your own database
  databaseProvider: keycloak.DatabaseProvider.fromDatabaseInfo({
    // Provide an RDS-compatible secret with credentials and connection
    // info (required)
    credentials: rdsDb.secret!,
    // Inform Keycloak of the database vendor (required)
    vendor: keycloak.KeycloakDatabaseVendor.MYSQL,
    // Add an ingress rule to the database security group (optional as long
    // as the Keycloak tasks can connect to the database)
    connectable: rdsDb,
  }),
  // Bring your own load balancer
  httpPortPublisher: keycloak.PortPublisher.addTarget({
    // Your load balancer listener
    listener,
    // Only publish certain paths
    conditions: [elbv2.ListenerCondition.pathPatterns([
      '/auth/*',
    ])],
    // Set your listener rule priority
    priority: 1000,
  }),
});
```

### Customize the container image

You may build and use a custom container image to add a custom theme, run scripts, add custom realm configs, or bundle your
own Keycloak SPIs.

```ts
new keycloak.KeycloakCluster(this, 'Keycloak', {
  keycloak: {
    image: ecs.ContainerImage.fromAsset(pathToDockerBuildContext, {
      buildArgs: {
        FROM: 'jboss/keycloak:12.0.2',
      },
    }),
  },
});
```

### Publish container ports through a Network Load Balancer

You may use the `PortPublisher` pattern to publish container ports through a Network Load Balancer.

```ts
new keycloak.KeycloakCluster(this, 'Keycloak', {
  // Publish the container's HTTP web port in a NLB on port 8080
  httpPortPublisher: keycloak.PortPublisher.nlb({
    port: 8080,
  }),
  // Publish the container's HTTPS port in an NLB on port 8443
  httpsPortPublisher: keycloak.PortPublisher.nlb({
    port: 8443,
    healthCheck: false,
  }),
  // Publish the Wildfly Admin Console on port 9990 (not recommended in
  // production)
  adminConsolePortPublisher: keycloak.PortPublisher.nlb({
    port: 9990,
    healthCheck: false,
  }),
});
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.184](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.182...v0.0.184) (2021-07-06)

### [0.0.183](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.182...v0.0.183) (2021-07-06)

### [0.0.182](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.181...v0.0.182) (2021-07-05)

### [0.0.181](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.180...v0.0.181) (2021-07-04)

### [0.0.180](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.179...v0.0.180) (2021-07-03)

### [0.0.179](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.178...v0.0.179) (2021-07-02)

### [0.0.178](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.177...v0.0.178) (2021-07-01)

### [0.0.177](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.176...v0.0.177) (2021-06-30)

### [0.0.176](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.175...v0.0.176) (2021-06-29)

### [0.0.175](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.174...v0.0.175) (2021-06-29)

### [0.0.174](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.173...v0.0.174) (2021-06-28)

### [0.0.173](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.172...v0.0.173) (2021-06-28)

### [0.0.172](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.171...v0.0.172) (2021-06-27)

### [0.0.171](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.170...v0.0.171) (2021-06-26)

### [0.0.170](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.169...v0.0.170) (2021-06-25)

### [0.0.169](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.168...v0.0.169) (2021-06-24)

### [0.0.168](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.167...v0.0.168) (2021-06-23)

### [0.0.167](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.166...v0.0.167) (2021-06-23)

### [0.0.166](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.165...v0.0.166) (2021-06-22)

### [0.0.165](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.164...v0.0.165) (2021-06-22)

### [0.0.164](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.163...v0.0.164) (2021-06-21)

### [0.0.163](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.162...v0.0.163) (2021-06-20)

### [0.0.162](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.161...v0.0.162) (2021-06-19)

### [0.0.161](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.160...v0.0.161) (2021-06-18)

### [0.0.160](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.159...v0.0.160) (2021-06-17)

### [0.0.159](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.158...v0.0.159) (2021-06-17)

### [0.0.158](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.157...v0.0.158) (2021-06-16)

### [0.0.157](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.156...v0.0.157) (2021-06-15)

### [0.0.156](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.155...v0.0.156) (2021-06-14)

### [0.0.155](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.154...v0.0.155) (2021-06-13)

### [0.0.154](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.153...v0.0.154) (2021-06-12)

### [0.0.153](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.152...v0.0.153) (2021-06-11)

### [0.0.152](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.151...v0.0.152) (2021-06-10)

### [0.0.151](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.150...v0.0.151) (2021-06-09)

### [0.0.150](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.149...v0.0.150) (2021-06-08)

### [0.0.149](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.148...v0.0.149) (2021-06-07)

### [0.0.148](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.147...v0.0.148) (2021-06-06)

### [0.0.147](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.146...v0.0.147) (2021-06-05)

### [0.0.146](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.145...v0.0.146) (2021-06-04)

### [0.0.145](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.144...v0.0.145) (2021-06-03)

### [0.0.144](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.143...v0.0.144) (2021-06-02)

### [0.0.143](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.142...v0.0.143) (2021-06-01)

### [0.0.142](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.141...v0.0.142) (2021-05-31)

### [0.0.141](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.140...v0.0.141) (2021-05-30)

### [0.0.140](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.139...v0.0.140) (2021-05-29)

### [0.0.139](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.138...v0.0.139) (2021-05-28)

### [0.0.138](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.137...v0.0.138) (2021-05-27)

### [0.0.137](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.136...v0.0.137) (2021-05-26)

### [0.0.136](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.135...v0.0.136) (2021-05-25)

### [0.0.135](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.134...v0.0.135) (2021-05-24)

### [0.0.134](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.133...v0.0.134) (2021-05-23)

### [0.0.133](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.132...v0.0.133) (2021-05-22)

### [0.0.132](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.131...v0.0.132) (2021-05-21)

### [0.0.131](https://github.com/misterjoshua/cdk-ecs-keycloak/compare/v0.0.130...v0.0.131) (2021-05-20)


src/keycloak-container-extension.ts
import * as ecs from '@aws-cdk/aws-ecs';
import * as elbv2 from '@aws-cdk/aws-elasticloadbalancingv2';
import * as logs from '@aws-cdk/aws-logs';
import * as secretsmanager from '@aws-cdk/aws-secretsmanager';
import * as cloudmap from '@aws-cdk/aws-servicediscovery';
import * as cdk from '@aws-cdk/core';

/**
 * The database vendor.
 */
export enum KeycloakDatabaseVendor {
  /** H2 In-memory Database (Warning: data deleted when task restarts.) */
  H2 = 'h2',
  /** MySQL */
  MYSQL = 'mysql',
  /** MariaDB */
  MARIADB = 'mariadb',
  /** MSSQL (not yet supported, please submit a PR) */
  MSSQL = 'mssql',
  /** Oracle database (not yet supported, please submit a PR) */
  ORACLE = 'oracle',
  /** Postgres (not yet supported, please submit a PR) */
  POSTGRES = 'postgres',
}

/**
 * Configuration for the Keycloak container.
 */
export interface KeycloakContainerExtensionProps {
  /**
   * Keycloak container image to use.
   * @default - use jboss/keycloak from docker hub.
   */
  readonly image?: ecs.ContainerImage;

  /**
   * A name for the container added to the task definition.
   * @default 'keycloak'
   */
  readonly containerName?: string;

  /**
   * Enable infinicache clustering.
   * @default - true if any cache owner count is greater than 1
   */
  readonly infinicacheClustering?: boolean;

  /**
   * The default number of distributed cache owners for each key.
   * @default 1
   */
  readonly cacheOwnersCount?: number;

  /**
   * The number of distributed cache owners for authentication sessions.
   * @default - same as `cacheOwnersCount`
   */
  readonly cacheOwnersAuthSessionsCount?: number;

  /**
   * Secrets manager secret containing the RDS database credentials and
   * connection information in JSON format.
   * @default - none
   */
  readonly databaseCredentials?: secretsmanager.ISecret;

  /**
   * Database name
   * @default 'keycloak'
   */
  readonly databaseName?: string;

  /**
   * Database schema
   * @default - for Postgresql, the default is 'public'
   */
  readonly databaseSchema?: string;

  /**
   * The database vendor.
   * @default KeycloakDatabaseVendor.H2
   */
  readonly databaseVendor?: KeycloakDatabaseVendor;

  /**
   * Default admin user. This user is created in the master realm if it doesn't exist.
   * @default 'admin'
   */
  readonly defaultAdminUser?: string;

  /**
   * Default admin user's password. This password is applied when the default admin user
   * is created.
   * @default 'admin'
   */
  readonly defaultAdminPassword?: string;

  /**
   * Memory limit of the keycloak task.
   * @default 1024
   */
  readonly memoryLimitMiB?: number;

  /**
   * Memory reservation size for the keycloak task.
   * @default - 80% of memoryLimitMiB
   */
  readonly memoryReservationMiB?: number;

  /**
   * Log driver for the task.
   * @default - cloudwatch with one month retention
   */
  readonly logging?: ecs.LogDriver;
}

/**
 * Adds a keycloak container to a task definition. To use ECS service discovery
 * to locate cluster members, you need to call `useCloudMapService` with the
 * CloudMap service so that we can configure the correct DNS query.
 */
export class KeycloakContainerExtension implements ecs.ITaskDefinitionExtension {
  /**
   * Name of the container added to the task definition.
   */
  public readonly containerName: string;

  /**
   * Name of the Keycloak database.
   */
  public readonly databaseName: string;

  /**
   * Database vendor.
   */
  public readonly databaseVendor: KeycloakDatabaseVendor;

  /**
   * The number of distributed cache owners for each key.
   */
  public readonly cacheOwnersCount: number;

  /**
   * The number of distributed auth session cache owners for each key.
   */
  public readonly cacheOwnersAuthSessionsCount: number;

  /**
   * The default admin user's name.
   */
  public readonly defaultAdminUser: string;

  /**
   * The default admin user password.
   */
  public readonly defaultAdminPassword: string;

  /**
   * True if infinicache clustering is enabled.
   */
  public readonly infinicacheClustering: boolean;

  /**
   * Web traffic port.
   */
  public readonly webPort: number = 8080;

  /**
   * Web traffic port with HTTPS
   */
  public readonly httpsWebPort: number = 8443;

  /**
   * Admin console port.
   */
  public readonly adminConsolePort: number = 9990;

  // Privates
  private readonly _memoryLimitMiB?: number;
  private readonly _memoryReservationMiB?: number;
  private readonly _logging: ecs.LogDriver;
  private readonly _databaseCredentials?: secretsmanager.ISecret;
  private readonly _image: ecs.ContainerImage;
  private _cloudMapService?: cloudmap.IService;
  private readonly _databaseSchema?: string;

  constructor(props?: KeycloakContainerExtensionProps) {
    this.cacheOwnersCount = props?.cacheOwnersCount ?? 1;
    this.cacheOwnersAuthSessionsCount = props?.cacheOwnersAuthSessionsCount ?? this.cacheOwnersCount;
    this._image = props?.image ?? ecs.ContainerImage.fromRegistry('jboss/keycloak');

    this.containerName = props?.containerName ?? 'keycloak';
    this.databaseVendor = props?.databaseVendor ?? KeycloakDatabaseVendor.H2;
    this.databaseName = props?.databaseName ?? 'keycloak';
    this._databaseSchema = props?.databaseSchema;
    this._databaseCredentials = props?.databaseCredentials;
    this.defaultAdminUser = props?.defaultAdminUser ?? 'admin';
    this.defaultAdminPassword = props?.defaultAdminPassword ?? 'admin';

    this._memoryLimitMiB = props?.memoryLimitMiB;
    this._memoryReservationMiB = props?.memoryReservationMiB;

    this._logging = props?.logging ?? ecs.LogDriver.awsLogs({
      streamPrefix: '/cdk-ecs-keycloak',
      logRetention: logs.RetentionDays.ONE_MONTH,
    });

    // Enable infinicache clustering by default if any cache owner count is greater than 1
    const defaultInfinicacheClustering = this.cacheOwnersCount > 1 || this.cacheOwnersAuthSessionsCount > 1;
    this.infinicacheClustering = props?.infinicacheClustering ?? defaultInfinicacheClustering;

    if (!this.infinicacheClustering) {
      if (this.cacheOwnersCount > 1 || this.cacheOwnersAuthSessionsCount > 1) {
        throw new Error('Infinicache clustering must be enabled for any cache owner count greater than 1');
      }
    }

    if (!isSupportedDatabaseVendor(this.databaseVendor)) {
      throw new Error(`The ${this.databaseVendor} engine is not yet tested and fully supported. Please submit a PR.`);
    }

    if (!this._databaseCredentials && this.databaseVendor !== KeycloakDatabaseVendor.H2) {
      throw new Error(`The ${this.databaseVendor} database vendor requires credentials`);
    }
  }

  /**
   * Inform Keycloak of a CloudMap service discovery mechanism.
   */
  useCloudMapService(serviceDiscovery: cloudmap.IService) {
    this._cloudMapService = serviceDiscovery;
  }

  /**
   * @inheritDoc
   */
  extend(taskDefinition: ecs.TaskDefinition): void {
    const keycloakSecrets: Record<string, ecs.Secret> = {};

    const databaseNameForVendor = this.databaseVendor != KeycloakDatabaseVendor.H2 ? this.databaseName : '';

    let keycloakMemoryLimit: number;
    if (this._memoryLimitMiB) {
      keycloakMemoryLimit = this._memoryLimitMiB;
    } else if (taskDefinition.isFargateCompatible && !this._memoryLimitMiB) {
      const cfnTaskDefinition = taskDefinition.node.defaultChild as ecs.CfnTaskDefinition;
      keycloakMemoryLimit = parseInt(cfnTaskDefinition.memory!);
    } else {
      keycloakMemoryLimit = 512;
    }

    // User-specified memory reservation, otherwise 80% of the memory limit.
    let keycloakMemoryReservation = this._memoryReservationMiB ?? Math.round(keycloakMemoryLimit * 0.8);

    if (this._databaseCredentials) {
      keycloakSecrets.DB_ADDR = ecs.Secret.fromSecretsManager(this._databaseCredentials, 'host');
      keycloakSecrets.DB_PORT = ecs.Secret.fromSecretsManager(this._databaseCredentials, 'port');
      keycloakSecrets.DB_USER = ecs.Secret.fromSecretsManager(this._databaseCredentials, 'username');
      keycloakSecrets.DB_PASSWORD = ecs.Secret.fromSecretsManager(this._databaseCredentials, 'password');
    }

    const keycloak = taskDefinition.addContainer(this.containerName, {
      image: this._image,
      environment: {
        KEYCLOAK_USER: this.defaultAdminUser,
        KEYCLOAK_PASSWORD: this.defaultAdminPassword,
        DB_VENDOR: this.databaseVendor,
        DB_NAME: databaseNameForVendor,
        DB_SCHEMA: this._databaseSchema ?? '',
        JGROUPS_DISCOVERY_PROTOCOL: cdk.Lazy.string({
          produce: () => this._getJGroupsDiscoveryProtocol(),
        }),
        JGROUPS_DISCOVERY_PROPERTIES: cdk.Lazy.string({
          produce: () => this._getJGroupsDiscoveryProperties(),
        }),
        CACHE_OWNERS_COUNT: this.cacheOwnersCount.toString(),
        CACHE_OWNERS_AUTH_SESSIONS_COUNT: this.cacheOwnersAuthSessionsCount.toString(),
        JDBC_PARAMS: 'useSSL=false',
        JAVA_OPTS: '-Djboss.bind.address.management=0.0.0.0',
      },
      secrets: keycloakSecrets,
      logging: this._logging,
      memoryLimitMiB: keycloakMemoryLimit,
      memoryReservationMiB: keycloakMemoryReservation,
    });


    keycloak.addPortMappings({ containerPort: this.webPort }); // Web port
    keycloak.addPortMappings({ containerPort: this.httpsWebPort }); // HTTPS web port
    keycloak.addPortMappings({ containerPort: this.adminConsolePort }); // Admin console port
    keycloak.addPortMappings({ containerPort: 7600 }); // jgroups-tcp
    keycloak.addPortMappings({ containerPort: 57600 }); // jgroups-tcp-fd
    keycloak.addPortMappings({ containerPort: 55200, protocol: ecs.Protocol.UDP }); // jgroups-udp
    keycloak.addPortMappings({ containerPort: 54200, protocol: ecs.Protocol.UDP }); // jgroups-udp-fd
  }

  /**
   * @internal
   */
  public _getJGroupsDiscoveryProtocol() {
    if (!this.infinicacheClustering) {
      return '';
    } else if (!this._cloudMapService) {
      return 'JDBC_PING';
    } else {
      return 'dns.DNS_PING';
    }
  }

  /**
   * @internal
   */
  public _getJGroupsDiscoveryProperties() {
    if (!(this.infinicacheClustering && this._cloudMapService)) {
      return '';
    }

    // Note: SRV-based discovery isn't enough to handle bridged-mode networking.
    // - Keycloak wants two ports for clustering in either stack mode
    // - CloudMap currently supports only one service registry per ecs service
    //
    // To the reader: Got any suggestions? Open a PR. I'd love to run this on
    // EC2 with bridged networking so that keycloak can be run in containers on
    // bursting instance types where vpc trunking is not available.
    return cdk.Fn.sub('dns_query=${ServiceName}.${ServiceNamespace},dns_record_type=${QueryType}', {
      ServiceName: this._cloudMapService.serviceName,
      ServiceNamespace: this._cloudMapService.namespace.namespaceName,
      QueryType: mapDnsRecordTypeToJGroup(this._cloudMapService.dnsRecordType),
    });
  }

  /**
   * Configure health checks on the target group.
   * @param targetGroup
   */
  public configureHealthCheck(targetGroup: elbv2.ApplicationTargetGroup) {
    targetGroup.configureHealthCheck({
      path: '/auth/realms/master',
      enabled: true,
    });
  }
}

/**
 * Checks if the given database vendor is supported by this construct.
 * @internal
 */
export function isSupportedDatabaseVendor(databaseVendor: KeycloakDatabaseVendor) {
  switch (databaseVendor) {
    case KeycloakDatabaseVendor.H2:
    case KeycloakDatabaseVendor.MYSQL:
    case KeycloakDatabaseVendor.POSTGRES:
      return true;

    default:
      return false;
  }
}

/**
 * Maps a cloudmap dns record type to a name recognized by the Keycloak container.
 * @internal
 */
export function mapDnsRecordTypeToJGroup(dnsRecordType: cloudmap.DnsRecordType): string {
  switch (dnsRecordType) {
    case cloudmap.DnsRecordType.A: return 'A';
    case cloudmap.DnsRecordType.SRV: return 'SRV';
    default:
      throw new Error(`Unsupported service discovery record type: ${dnsRecordType}`);
  }
}